### PR TITLE
[Core] Add backward compatibility with vllm

### PR DESCRIPTION
**Pull request by Shaoting-Feng**
Make it compatible with `mm_hashes` besides the latest `mm_features`.

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to LMCache! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core LMCache logic (e.g., <code>LMCacheEngine</code>, <code>Backend</code> etc.)</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient unit tests to ensure the change is stay correct and robust. The unit and integration tests will always run and our comprehensive test will be triggered after the "full" label is tagged onto a PR.</li>
</ul>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of KuntaiDu, ApostaC or YaoJiayi.

</details>

### CODE REVIEW

**Comment by Shaoting-Feng** [comment]
Ran the same tests as in https://github.com/LMCache/LMCache/pull/1582 and they still work.

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 1]
```diff
@@ -1055,23 +1058,27 @@ def get_num_new_matched_tokens(
         token_ids = request.prompt_token_ids
 
         # If the request has multimodal hashes, apply them to the token ids
-        if request.mm_features:
+        if request.mm_features or request.mm_hashes:
```
Comment (ApostaC) [inline lmcache/integration/vllm/vllm_v1_adapter.py]: Can we have a helper function called `extract_mm_features` that returns `Optional[Tuple[mm_hashes, mm_positions]]`?

So the code will be like:
```python
if mm_features := extract_mm_features(request):
    token_ids = torch.tensor(request.prompt_token_ids)
    mm_hashes, mm_positions = mm_features
    apply_mm_hashes_to_token_ids(.....)
```

**Review by ApostaC — COMMENTED** [review]
Quick idea to simplify the code a bit

File: lmcache/integration/vllm/vllm_v1_adapter.py (hunk 1 above)
Comment (Shaoting-Feng) [reply lmcache/integration/vllm/vllm_v1_adapter.py]: Thanks for this suggestion!

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 2]
```diff
@@ -1055,23 +1058,27 @@ def get_num_new_matched_tokens(
...
+                apply_mm_hashes_to_token_ids(
+                    token_ids,
+                    list(mm_hashes),
+                    list(mm_positions),
+                )
+            elif request.mm_hashes:
+                apply_mm_hashes_to_token_ids(
+                    token_ids, request.mm_hashes, request.mm_positions
+                )
             token_ids = token_ids.tolist()
```
[the remark this answers is not in the document]
Comment (Shaoting-Feng) [reply lmcache/integration/vllm/vllm_v1_adapter.py]: Yihua proposed better solution.

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 3]
```diff
@@ -1055,23 +1052,17 @@ def get_num_new_matched_tokens(
         token_ids = request.prompt_token_ids
 
         # If the request has multimodal hashes, apply them to the token ids
-        if request.mm_features:
+        mm_features = extract_mm_features(request)
+        if mm_features[0] != [] and mm_features[1] != []:
```
Comment (ApostaC) [inline lmcache/integration/vllm/vllm_v1_adapter.py]:
```suggestion
        mm_hashes, mm_positions = extract_mm_features(request)
        if mm_hashes and mm_positions:
```

File: lmcache/integration/vllm/vllm_v1_adapter.py [hunk 4]
```diff
@@ -1055,23 +1052,17 @@ def get_num_new_matched_tokens(
...
             token_ids = torch.tensor(request.prompt_token_ids)
-            mm_hashes, mm_positions = zip(
-                *((f.identifier, f.mm_position) for f in request.mm_features)
-            )
-            apply_mm_hashes_to_token_ids(
-                token_ids,
-                list(mm_hashes),
-                list(mm_positions),
-            )
+            mm_hashes, mm_positions = mm_features
```
Comment (ApostaC) [inline lmcache/integration/vllm/vllm_v1_adapter.py]:
```suggestion
```

**Review by ApostaC — APPROVED** [review]
otherwise LGTM

**Review by YaoJiayi — APPROVED** [review]
